Fix ShouldRestart for on-failure handle

WeiZhang555 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"
-->

Currently if you restart docker daemon, all the containers with restart
policy `on-failure` regardless of its `RestartCount` will be started, this
will make daemon cost more extra time for restart.

This commit will stop these containers to do unnecessary start on
daemon's restart.

How to reproduce:
1. start a container with restart policy `on-failure:3`
`$docker run -tid --restart on-failure:3 busybox sh -c "sleep 1; exit 127"`
2. wait it to fail and restart 3 times, make sure it stops and `docker ps` can't see it.
3. Restart docker daemon.
4. (DO THIS QUICKLY) `docker ps` shows that the stopped container is started, and fails, and restart 3 times again. 

This is unnecessary because it already failed 3 times last time daemon is running, and starting this container will apparently lower the restart speed of daemon.

---

_Edit_: make some small refactor
Merge `Container.ShouldRestart()` and `containerMonitor.shouldRestart()`, so
they can share same logic. Besides that, some duplicated fields of
containerMonitor are removed.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

thaJeztah commented:
Should we have a test for this?

change LGTM, although the code is becoming a bit complex to read, possibly splitting it up at some point would be good for readability

WeiZhang555 commented:
This test case will need to restart docker daemon, I'm not familiar with that part, could someone give me some instructions? 

And I wonder it will become a flaky test, because it need to capture the container's restarting/running state.

thaJeztah commented:
ping @vdemeester wdyt, is it worth a test, or are we okay without?

WeiZhang555 commented:
@thaJeztah I figure out how to make the test case now, I think it'll work :smile: 

/cc @vdemeester @cpuguy83 @calavera

cpuguy83 commented on container/container.go:
```diff
@@ -505,7 +505,8 @@ func copyEscapable(dst io.Writer, src io.ReadCloser, keys []byte) (written int64
 	return container.HostConfig.RestartPolicy.Name == "always" ||
 		(container.HostConfig.RestartPolicy.Name == "unless-stopped" && !container.HasBeenManuallyStopped) ||
-		(container.HostConfig.RestartPolicy.Name == "on-failure" && container.ExitCode != 0)
+		(container.HostConfig.RestartPolicy.Name == "on-failure" && container.ExitCode != 0 &&
```
Can we have `container/monitor.go` use this fn as well? It looks like it's defining it's own rules, which are exactly the same.

cpuguy83 commented on integration-cli/docker_cli_daemon_test.go:
```diff
@@ -150,6 +150,45 @@ func (s *DockerDaemonSuite) TestDaemonRestartUnlessStopped(c *check.C) {
+
+	// wait test1 to stop
+	var stopped bool
+	for i := 0; i < 6; i++ {
```
How about `waitInspect` instead of this loop?

WeiZhang555 replied:
Good idea, I will update as per your suggestion :+1: :smile:

WeiZhang555 commented:
@cpuguy83 I made an implementation for having `container/monitor.go` use `Container.ShouldRestart()`, it's a little complicated now and I'm worrying it may bring race conditions.
But in general it should make sense, let's see whether Janky is happy with this change.

WeiZhang555 commented:
Double checked this change, I believe that it should be right, Janky's test result also supported me.

/cc @vdemeester @calavera

thaJeztah commented:
WindowsTP4 timed out after 2 hours :cry: restarting

WeiZhang555 commented:
@thaJeztah Looks like 2 hours is definitely insufficient for Windows now :crystal_ball:

thaJeztah commented:
@WeiZhang555 yeah, I brought it up yesterday, but we think https://github.com/docker/docker/pull/21017 will be a better solution than raising the timeout limit. Perhaps I should discuss it again, even if just temporarily raising the limit :cry:

thaJeztah commented:
ping @WeiZhang555 sorry, needs a rebase now :cry:

WeiZhang555 commented:
@thaJeztah Rebased, thank you for your reminder!

/cc @cpuguy83

cpuguy83 commented:
LGTM, thanks!

vdemeester commented on container/container.go:
```diff
@@ -530,9 +531,26 @@ func copyEscapable(dst io.Writer, src io.ReadCloser, keys []byte) (written int64
+	case rp.IsAlways():
+		return true
+	case rp.IsUnlessStopped():
+		if !container.HasBeenManuallyStopped {
```
It's a nit but I would have inverted the condition, so it's `if container.HasBeenManuallyStopped {` instead (feels easier to read).

WeiZhang555 replied:
OK

vdemeester commented on container/monitor.go:
```diff
@@ -249,13 +241,6 @@ func (m *containerMonitor) resetMonitor(successful bool) {
 		m.timeIncrement *= 2
 	}
-
-	// the container exited successfully so we need to reset the failure counter
```
@WeiZhang555 I'm not sure why we are removing this ?

WeiZhang555 replied:
`failureCount` is only used in 
`func (m *containerMonitor) shouldRestart(exitCode int) bool` for calculating the Restart Condition, I have merged `container.ShouldRestart` and `containerMonitor.shouldRestart` as @cpuguy83 suggested, and `failureCount` can be replaced with `container.RestartCount`

We don't need it, hence removed.

vdemeester replied:
oh you're right :smile_cat:

thaJeztah commented:
ping @vdemeester PR was updated, ptal

WeiZhang555 commented:
@thaJeztah Thank you :stuck_out_tongue_winking_eye:

vdemeester commented:
LGTM 🐼

thaJeztah commented:
@WeiZhang555 oh boy, looks like it needs a rebase now. Can you rebase, then I think we can merge (I see two LGTM's)

WeiZhang555 commented:
@thaJeztah Yeah, the merge of containerd conflicts a lot with this, I will make some changes of code to adapt new code. 

Will post more updates soon. :smile:

thaJeztah commented:
Yes, I feared that would be the case; sorry for that :blush:

WeiZhang555 commented:
@thaJeztah Never mind, this change is still necessary as I can see, just need some time to do more investigate, and also another code review :cry:

WeiZhang555 committed:
Fix ShouldRestart for on-failure handle

Currently if you restart docker daemon, all the containers with restart
policy `on-failure` regardless of its `RestartCount` will be started,
this will make daemon cost more extra time for restart.

This commit will stop these containers to do unnecessary start on
daemon's restart.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
@cpuguy83 because of the change brought by `containerd`, it's hard to merge ShouldRestart() now. So I removed the refactor and only keeps the bug fix.  

I can try to do the refactor after I get more familiar with new code if you don't mind.

Also ping @vdemeester @thaJeztah for another code review. :smile:

vdemeester commented:
Re-LGTM if it's :green_book:

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -310,6 +310,8 @@ func (daemon *Daemon) restore() error {
 			}
 
+			// reset container restart count
+			c.RestartManager(true)
```
Do we really want to reset all container's restart policies here?

WeiZhang555 replied:
RestartManager decides whether the container should be restarted based on its `failureCount`,  If we don't reset container's restart count, a container with `on-failure:3` may have a `RestartCount` 5. e.g.
1. container failed 2 times, `RestartCount`=2
2. daemon shutdown.
3. daemon reboot, *RestartCount not reset*, container failed and restart 3 times, RestartCount turns out to be 2+3=5.  

This is incorrect.

cpuguy83 replied:
So this is a problem with some internal counter that's not persisted?

WeiZhang555 replied:
This is because we used two variables to calculate ShouldRestart: failureCount and RestartPolicy.

OK, I'd better merge these two params again, have had done it before, I'll try to migrate that part into this.

thaJeztah commented:
ping @cpuguy83 ptal, what's the status here?


@WeiZhang555 looks like it needs a rebase :cry:

cpuguy83 commented:
I think we're just waiting on the changes mentioned above.

WeiZhang555 commented:
Yes, sorry for my delay. I'll be on travel these days and can't get reach of my computer.I'll post more updates once I come back in 3 or 4 days.

thaJeztah commented:
@WeiZhang555 no worries; we were just going through some old PR's, thanks in advance!

WeiZhang555 commented:
I'm back again, the code is updated based on previous comment, please help take a look. @cpuguy83

cpuguy83 commented on restartmanager/restartmanager.go:
```diff
@@ -72,11 +73,13 @@ func (rm *restartManager) ShouldRestart(exitCode uint32) (bool, chan error, erro
 	case rm.policy.IsOnFailure():
 		// the default value of 0 for MaximumRetryCount means that we will not enforce a maximum count
-		if max := rm.policy.MaximumRetryCount; max == 0 || rm.failureCount <= max {
+		if max := rm.policy.MaximumRetryCount; max == 0 || rm.restartCount < max {
```
How come you changed this to just `<`?

WeiZhang555 replied:
I replaced the "failureCount" in `restartManager` with `restartCount` for reusing the `restartManager.ShouldRestart()` in `container.ShouldRestart()`, when using failureCount, we're doing this:
1. failureCount++
2. Judge if failureCount <= max
After replaced with `restartCount`, we're doing:
1. Judge if restartCount < max
2. if need restart, restartCount++

And they have different meaning, `restartCount`means how many times container has been restarted before, it must be less than MaximumRetryCount, if the limit is reached, we shall not restart it again.

cpuguy83 commented:
1 minor question, otherwise LGTM

WeiZhang555 commented:
ping @cpuguy83

cpuguy83 commented:
LGTM but sadly needs a rebase again :(

WeiZhang555 commented:
Rebased. Waiting Janky turn to green :smile:

vdemeester commented:
LGTM :+1:

thaJeztah commented:
Thanks @WeiZhang555!

WeiZhang555 commented:
:laugh:

thaJeztah commented:
Added "changelog" label; May be a nice optimization to mention in the change logs
